util/mathutil/exponential_average.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mathutil
+
+import "sync"
+
+// ExponentialMovingAverage is an exponential moving average measurement implementation.
+type ExponentialMovingAverage struct {
+	value        float64
+	sum          float64
+	factor       float64
+	warmupWindow int
+	count        int
+
+	mu sync.RWMutex
+}
+
+// NewExponentialMovingAverage will create a new ExponentialMovingAverage
+func NewExponentialMovingAverage(
+	factor float64,
+	warmupWindow int,
+) *ExponentialMovingAverage {
+	if factor >= 1 || factor == 0 {
+		panic("factor must be (0, 1)")
+	}
+	return &ExponentialMovingAverage{
+		factor:       factor,
+		warmupWindow: warmupWindow,
+	}
+}
+
+// Add a single sample and update the internal state.
+func (m *ExponentialMovingAverage) Add(value float64) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.count < m.warmupWindow {
+		m.count++
+		m.sum += value
+		m.value = m.sum / float64(m.count)
+	} else {
+		m.value = m.value*m.factor + value*(1-m.factor)
```

Comment:
```suggestion
		m.value = m.value*(1-m.factor) + value*m.factor
```

I prefer a more common format that factor is applied on new values rather than old averages

https://tttapa.github.io/Pages/Mathematics/Systems-and-Control-Theory/Digital-filters/Exponential%20Moving%20Average/Exponential-Moving-Average.html